testlapack/dlantr.go

```diff
@@ -0,0 +1,77 @@
+package testlapack
+
+import (
+	"math"
+	"math/rand"
+	"testing"
+
+	"github.com/gonum/blas"
+	"github.com/gonum/lapack"
+)
+
+type Dlantrer interface {
+	Dlanger
+	Dlantr(norm lapack.MatrixNorm, uplo blas.Uplo, diag blas.Diag, m, n int, a []float64, lda int, work []float64) float64
+}
+
+func DlantrTest(t *testing.T, impl Dlantrer) {
+	for _, norm := range []lapack.MatrixNorm{lapack.MaxAbs, lapack.MaxColumnSum, lapack.MaxRowSum, lapack.NormFrob} {
+		for _, diag := range []blas.Diag{blas.NonUnit, blas.Unit} {
+			for _, uplo := range []blas.Uplo{blas.Lower, blas.Upper} {
+				for _, test := range []struct {
+					m, n, lda int
+				}{
+					{3, 3, 0},
+					{3, 5, 0},
+					{10, 5, 0},
+
+					{5, 5, 11},
+					{5, 10, 11},
+					{10, 5, 11},
+				} {
+					// Do a couple of random trials since the values change.
+					for i := 0; i < 1; i++ {
+						m := test.m
+						n := test.n
+						lda := test.lda
+						if lda == 0 {
+							lda = n
+						}
+						a := make([]float64, m*lda)
+						for i := range a {
+							a[i] = rand.NormFloat64()
```

Comment:
Have at least one round where a[i] is set to i - this makes it easier to see what is going wrong.